designs/odcr.md

```diff
@@ -0,0 +1,204 @@
+
+## Proposed Solution
+
+### Supporting associating Capacity Reservations to EC2NodeClass
+
+- Add a new field under `spec` for `capacityReservationSelectorTerms` to `EC2NodeClass` for defining which Capacity Reservation to be used for a specific `EC2NodeClass`
+- Add a new field under `status` for the found Capacity Reservations by the `spec.capacityReservationSelectorTerms` for the `EC2NodeClass`
+
+```yaml
+apiVersion: karpenter.k8s.aws/v1beta1
+kind: EC2NodeClass
+metadata:
+  name: example-node-class
+spec:
+  capacityReservationSelectorTerms:
+    - # The Availability Zone of the Capacity Reservation
+      availabilityZone: String | None
+      # The platform of operating system for which the Capacity Reservation reserves capacity
+      id: String | None
+      # The type of operating system for which the Capacity Reservation reserves capacity
+      instancePlatform: String | None
+      # The type of operating system for which the Capacity Reservation reserves capacity
+      instanceType: String | None
+      # The ID of the Amazon Web Services account that owns the Capacity Reservation
+      ownerId: String | None
+      # Tags is a map of key/value tags used to select subnets
+      # Specifying '*' for a value selects all values for a given tag key.
+      tags: Map | None
+      # Indicates the tenancy of the Capacity Reservation.
+      # A Capacity Reservation can have one of the following tenancy 'default' or 'dedicated':
+      #   default - The Capacity Reservation is created on hardware that is shared with other Amazon Web Services accounts.
+      #   dedicated - The Capacity Reservation is created on single-tenant hardware that is dedicated to a single Amazon Web Services account.
+      tenancy: String | None
+status:
+  capacityReservations:
+    - # AvailabilityZone of the Capacity Reservation
+      availabilityZone: String
+      # Available Instance Count of the Capacity Reservation
+      availableInstanceCount: Integer
+      # ID of the Capacity Reservation
+      id: String
+      # InstanceType of the Capacity Reservation
+      instanceType: String
+      # Total Instance Count of the Capacity Reservation
+      totalInstanceCount: Integer
+```
+
+This follows closely (does not implement all fields) how EC2 [DescribeCapacityReservations](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_DescribeCapacityReservations.html) can filter.
+
+Karpenter will perform validation against the spec to ensure there isn't any violation prior to creating the LaunchTemplates.
+
+### Supporting new capacity-type "capacity-reservation" for NodePool
+
+Adding a new `karpenter.sh/capacity-type: capacity-reservation` allows us to have a EC2NodeClass that does not automatically fallback to `on-demand` if `capacity-reservation` is not available.
```

Comment:
I would strongly suggest you incorporate the reservation type into this so you can differentiate an ODCR from a Capacity Block.

Are we sure capacity-type is the right way to model this? Right now, capacity-type aligns with EC2 instance metadata for market type. This would be a notable divergence from that existing philosophy. I'm wondering if we can't add a new top-level requirement instead of overloading capacity-type?